extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectDirectory.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.guacamole.auth.quickconnect;
+
+import java.util.Collection;
+import org.apache.guacamole.GuacamoleException;
+import org.apache.guacamole.auth.quickconnect.utility.QCParser;
+import org.apache.guacamole.net.auth.ConnectionGroup;
+import org.apache.guacamole.net.auth.simple.SimpleConnectionDirectory;
+import org.apache.guacamole.net.auth.Connection;
+import org.apache.guacamole.protocol.GuacamoleConfiguration;
+
+/**
+ * Implementation of the Connection Directory, stored
+ * completely in-memory.
+ */
+public class QuickConnectDirectory extends SimpleConnectionDirectory {
+
+    /**
+     * The unique identifier of the root connection group.
+     */
+    private static final String ROOT_IDENTIFIER = "ROOT";
+
+    /**
+     * The root connection group for this directory.
+     */
+    private final QuickConnectConnectionGroup rootGroup;
+
+    /**
+     * The internal counter for connection IDs.
+     */
+    private int CONNECTION_ID = 0;
+
+    /**
+     * Creates a new QuickConnectDirectory which provides access to the
+     * connections contained within the given Map.
+     *
+     * @param connections
+     *     A Collection of all connections that should be present in this
+     *     connection directory.
+     * @param rootGroup
+     *     A group that should be at the base of this directory.
+     */
+    public QuickConnectDirectory(Collection<Connection> connections, ConnectionGroup rootGroup) {
+        super(connections);
+        this.rootGroup = (QuickConnectConnectionGroup)rootGroup;
+    }
+
+    /**
+     * Returns the current counter and then increments it.
+     *
+     * @return
+     *     An Integer representing the next available connection
+     *     ID to get used when adding connections.
+     */
+    private Integer getNextConnectionID() {
+        return CONNECTION_ID++;
+    }
+
+    @Override
+    public void add(Connection object) throws GuacamoleException {
+
+        put(new QuickConnection(object));
```

Comment:
Out of curiosity, is it an intended feature that these connections persist in memory until the end of the session? Or is that a side effect of the implementation, with the only goal being to provide some way to connect in an ad hoc manner?

Reply:
It was probably initially a side-effect, but one that I liked so I made it intentional.  How's that?! :-)

Obviously the main goal is to connect in an ad-hoc manner; however, I think it's also useful to be able to quickly retrieve those connections in the same session - i.e., I get kicked out of my session due to a network issue and want to quickly reconnect without typing the entire connection string, again.

Also, in one iteration of the implementation I had a little box where it looked for other available connection directories, and, if you had permission, allowed you to use the ad-hoc box to create a connection within the directory.  I don't think I kept it, here - needs a little more work - but it might be something that gets tacked (back) on in the future.